src/PrizeVault.sol

```diff
@@ -528,16 +538,27 @@ contract PrizeVault is TwabERC20, Claimable, IERC4626, ILiquidationSource, Ownab
             _yieldFee = (_amountOut * FEE_PRECISION) / (FEE_PRECISION - _yieldFeePercentage) - _amountOut;
         }
 
-        // Ensure total withdrawn does not exceed the available yield balance:
-        uint256 _totalToWithdraw = _amountOut + _yieldFee;
-        if (_totalToWithdraw > _availableYieldBalance) {
-            revert LiquidationExceedsAvailable(_totalToWithdraw, _availableYieldBalance);
+        // Ensure total liquidation does not exceed the available yield balance:
+        uint256 _totalLiquidation = _amountOut + _yieldFee;
+        if (_totalLiquidation > _availableYield) {
+            revert LiquidationExceedsAvailable(_totalLiquidation, _availableYield);
         }
 
-        _withdraw(address(this), _totalToWithdraw);
-        _asset.transfer(_receiver, _amountOut);
-        if (_yieldFee > 0) {
-            _asset.transfer(_yieldFeeRecipient, _yieldFee);
+        if (_tokenOut == address(_asset)) {
+            // Withdraw assets:
+            _withdraw(address(this), _totalLiquidation);
+            _asset.transfer(_receiver, _amountOut);
+            if (_yieldFee > 0) {
+                _asset.transfer(_yieldFeeRecipient, _yieldFee);
+            }
+        } else if (_tokenOut == address(this)) {
+            // Mint shares:
+            _mint(_receiver, _amountOut);
+            if (_yieldFee > 0) {
+                _mint(_yieldFeeRecipient, _yieldFee);
+            }
+        } else {
+            revert LiquidationTokenOutNotSupported(_tokenOut);
```

Comment:
Interesting. It's cool to see this make use of the tokenOut.

Could use some tweaks tho; this code forces the yield fee recipient to take fees in whatever token the liquidator wants; not what the recipient wants.

It also adds additional gas costs. For the withdrawal, the code could withdraw directly to the receiver if it subtracted the fee from the total liquidation. This would save 2 token transfers.  For the mint, it would eliminate 1 mint call.

I think this is an interesting approach, but needs some changes:

- Track the accrued yield fee balance in storage. Withdraw directly to the receiver; mint once.
- Allow the fee recipient to mint/withdraw their fee, as before.

Reply:
Commit addressing these issues: https://github.com/GenerationSoftware/pt-v5-vault/pull/75/commits/8dbcc07e87b9f9794a5095f293bbb28b88b309cf

The yield fee now accrues as an internal share balance (`yieldFeeBalance`) that is accounted for in the `totalSupply` so there are no edge cases when it comes to redeem them for actual shares. The yield fee recipient can then call `claimYieldFeeShares` to convert that internal balance to actual shares they control and can withdraw/redeem through the ERC4626 functions.

I started adding two extra `mint` and `withdraw` functions to let them withdraw assets in one step, but it felt odd to duplicate the same logic that is in the existing `redeem` and `withdraw` functions without having it go through the internal `_burnAndWithdraw` function, so I left it as the simple `claimYieldFeeShares` function which just acts as a conversion from internal share balance to TwabController share balance.